### PR TITLE
Hooks results db persist

### PR DESCRIPTION
**Pull request by nopcoder**

### CODE REVIEW

File: actions/service.go [hunk 1]
```diff
@@ -17,31 +17,62 @@ type Service struct {
...
-	runID := NewRunID(event.EventTime)
-	tasks, err := s.allocateTasks(runID, actions)
+	tasks, err := s.allocateTasks(event.RunID, actions)
 	if err != nil {
 		return nil
 	}
-	return s.runTasks(ctx, tasks, event, deps)
+
+	runErr := s.runTasks(ctx, tasks, event, deps)
+	endTime := time.Now()
```
Comment (itaiad200) [inline actions/service.go:68]: Minor: max on the tasks EndTime?

File: actions/service.go [hunk 2]
```diff
@@ -74,24 +105,60 @@ func (s *Service) allocateTasks(runID string, actions []*Action) ([]*Task, error
...
-	runID := t.UTC().Format(time.RFC3339) + "_" + uid
-	return runID
+func (s *Service) insertRunInformation(ctx context.Context, event Event, tasks []*Task, endTime time.Time, runErr error) error {
+	_, err := s.DB.Transact(func(tx db.Tx) (interface{}, error) {
+		var err error
+		// insert run information
+		runPassed := runErr == nil
+		_, err = tx.Exec(`INSERT INTO actions_runs(repository_id, run_id, event_type, start_time, end_time, branch_id, source_ref, commit_id, passed)
+			VALUES ($1,$2,$3,$4,$5,$6,$7,'',$8)`,
+			event.RepositoryID, event.RunID.String(), event.EventType, event.EventTime, endTime, event.BranchID, event.SourceRef, runPassed)
```
Comment (itaiad200) [inline actions/service.go:134]: `event.EventTime` is the `StartTime` of the `Run` but not the `Task`s?

File: actions/event.go [hunk 3]
```diff
@@ -19,7 +21,7 @@ type Deps struct {
 }
 
 type Event struct {
-	EventID       uuid.UUID
+	RunID         uuid.UUID
 	EventType     EventType
```
Comment (itaiad200) [inline actions/event.go:25]:
```suggestion
	Type     EventType
```

File: actions/event.go [hunk 4]
```diff
@@ -19,7 +21,7 @@ type Deps struct {
...
 	EventTime     time.Time
```
Comment (itaiad200) [inline actions/event.go:26]:
```suggestion
	Time     time.Time
```

File: actions/service.go [hunk 5]
```diff
@@ -74,24 +105,60 @@ func (s *Service) allocateTasks(runID string, actions []*Action) ([]*Task, error
...
+		if err != nil {
+			return nil, fmt.Errorf("insert run information: %w", err)
+		}
+
+		// insert each task information
+		for _, task := range tasks {
+			taskPassed := runErr == nil
```
Comment (itaiad200) [inline actions/service.go:141]: `runErr` is global run error, not the individual task error.

File: graveler/ref/branch_locker_test.go [hunk 6]
```diff
@@ -200,7 +200,6 @@ func panicOnMetadataUpdate(bl *ref.BranchLocker) {
 		ctx := context.Background()
 		_, _ = bl.MetadataUpdater(ctx, "branch-locker", testutil.DefaultBranchID, func() (interface{}, error) {
 			panic("metadata updater")
-			return nil, nil
```
Comment (itaiad200) [inline graveler/ref/branch_locker_test.go:203]: Revert.

**Review by itaiad200 — CHANGES_REQUESTED** [review]

File: swagger.yml [hunk 7]
```diff
@@ -1813,7 +1799,12 @@ paths:
       - in: query
         name: type
         type: string
-        <<: *DIFF_TYPE
+      - in: query
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply swagger.yml:1802]: lint warning we had with swagger

File: actions/service.go (hunk 5 above)
Comment (nopcoder) [reply actions/service.go:141]: good catch, missed this one!

File: graveler/ref/branch_locker_test.go (hunk 6 above)
Comment (nopcoder) [reply graveler/ref/branch_locker_test.go:203]: go vet - unreachable code

File: actions/action.go [hunk 8]
```diff
@@ -137,16 +137,30 @@ func LoadActions(ctx context.Context, source Source) ([]*Action, error) {
...
 		return nil, err
 	}
+	if err := validateActions(actions); err != nil {
+		return nil, err
+	}
 	return actions, nil
 }
 
+// validateActions verify we do not two actions with the same name
+func validateActions(actions []*Action) error {
```
Comment (itaiad200) [inline actions/action.go:153]: Don't forget the validation about a single hook per action.

File: actions/service.go [hunk 9]
```diff
@@ -27,95 +27,71 @@ type Task struct {
 }
 
 type RunResult struct {
-	RunID      string
-	BranchID   string
-	OnRef      string
```
Comment (itaiad200) [inline actions/service.go]: Why removed?

File: actions/service.go [hunk 10]
```diff
@@ -27,95 +27,71 @@ type Task struct {
...
 }
-func (s *Service) Run(ctx context.Context, event Event, deps Deps) error {
+func (s *Service) Run(ctx context.Context, event Event, deps Deps) (string, error) {
+	var runID string
 	// load relevant actions
-	actions, err := s.loadMatchedActions(ctx, deps.Source, MatchSpec{EventType: event.EventType, Branch: event.BranchID})
+	actions, err := s.loadMatchedActions(ctx, deps.Source, MatchSpec{EventType: event.Type, Branch: event.BranchID})
 	if err != nil || len(actions) == 0 {
-		return err
+		return runID, err
```
Comment (itaiad200) [inline actions/service.go:78]:
```suggestion
		return "", err
```

File: actions/service.go [hunk 11]
```diff
@@ -27,95 +27,71 @@ type Task struct {
...
 	}
 
 	// allocate and run hooks
-	tasks, err := s.allocateTasks(event.RunID, actions)
+	runID = NewRunID()
+	tasks, err := s.allocateTasks(runID, actions)
 	if err != nil {
-		return nil
+		return runID, nil
```
Comment (itaiad200) [inline actions/service.go:85]:
```suggestion
		return runID, err
```

File: swagger.yml [hunk 12]
```diff
@@ -2284,10 +2283,10 @@ paths:
           name: run_id
           required: true
           type: string
-        - in: query
-          name: before
-          type: string
-          format: date-time
+#        - in: query
```
Comment (itaiad200) [inline swagger.yml:2286]: No `after` instead?

File: api/controller.go [hunk 13]
```diff
@@ -72,7 +72,7 @@ type actionsHandler interface {
 	GetRunResult(ctx context.Context, repositoryID string, runID string) (*actions.RunResult, error)
 	GetTaskResult(ctx context.Context, repositoryID string, runID string, actionName string, hookID string) (*actions.TaskResult, error)
 	ListRuns(ctx context.Context, repositoryID string, afterRunID string, branchID *string) (actions.RunResultIterator, error)
-	ListRunTasks(ctx context.Context, repositoryID string, afterRunID string) (actions.TaskResultIterator, error)
+	ListRunTasks(ctx context.Context, repositoryID string, runID string) (actions.TaskResultIterator, error)
```
Comment (itaiad200) [inline api/controller.go:75]: Do we assume no pagination needed?

File: actions/service.go [hunk 14]
```diff
@@ -244,10 +245,12 @@ func (s *Service) GetTaskResult(ctx context.Context, repositoryID string, runID
 	return res.(*TaskResult), nil
 }
 
-func (s *Service) ListRuns(ctx context.Context, repositoryID string, afterRunID string, branchID *string) (RunResultIterator, error) {
-	return nil, ErrNotFound
+func (s *Service) ListRuns(ctx context.Context, repositoryID string, fromRunID string, branchID *string) (RunResultIterator, error) {
```
Comment (itaiad200) [inline actions/service.go:248]: Why is requested amount not propagated to the service?

File: actions/service.go (hunk 9 above)
Comment (nopcoder) [reply actions/service.go]: Added as SourceRef - let me know if I missed something else, there was no other OnRef on the original event.

File: swagger.yml (hunk 12 above)
Comment (nopcoder) [reply swagger.yml:2286]: The one I wanted to discuss. In order to have pagination on a specific run's tasks, we need to have the combination of action name and hook it passed as a token or have run_id/action_name/hook_id and have pagination cross runs.

File: actions/action.go (hunk 8 above)
Comment (nopcoder) [reply actions/action.go:153]: I wish to remove this restriction soon, just need to address the behavior we like

File: actions/action.go [hunk 15]
```diff
@@ -50,6 +51,9 @@ func (a *Action) Validate() error {
 	if a.Name == "" {
 		return fmt.Errorf("%w 'name' is required", ErrInvalidAction)
```
Comment (itaiad200) [inline actions/action.go:52]:
```suggestion
		return fmt.Errorf("'name' is required: %w", ErrInvalidAction)
```

File: actions/action.go [hunk 16]
```diff
@@ -50,6 +51,9 @@ func (a *Action) Validate() error {
...
 	}
+	if !reName.MatchString(a.Name) {
+		return fmt.Errorf("%w 'name' is not valid", ErrInvalidAction)
```
Comment (itaiad200) [inline actions/action.go:55]:
```suggestion
		return fmt.Errorf("'name' is not valid: %w", ErrInvalidAction)
```

File: actions/action.go [hunk 17]
```diff
@@ -50,6 +51,9 @@ func (a *Action) Validate() error {
...
+	}
 	if a.On.PreMerge == nil && a.On.PreCommit == nil {
 		return fmt.Errorf("%w 'on' is required", ErrInvalidAction)
```
Comment (itaiad200) [inline actions/action.go:58]:
```suggestion
		return fmt.Errorf("'on' is required: %w", ErrInvalidAction)
```

File: actions/db_run_result_iterator.go [hunk 18]
```diff
@@ -57,19 +57,19 @@ func (it *DBRunResultIterator) maybeFetch() {
 	if len(it.buf) > 0 {
 		return
 	}
-
+	if it.state == iteratorStateInit {
```
Comment (itaiad200) [inline actions/db_run_result_iterator.go:60]: Why do you need Init & Query states?
I understand where `iteratorStateDone` is being taking care of, but I don't see the difference between the 2 states above, now that lines 69-72 are deleted.

File: actions/service.go [hunk 19]
```diff
@@ -70,24 +76,26 @@ func NewService(db db.Database) *Service {
 		DB: db,
 	}
 }
+
+// Run load and run actions based on the event information. Always returns run id in order to track back hooks errors
+//
```
Comment (itaiad200) [inline actions/service.go:81]: dd

File: actions/service.go [hunk 20]
```diff
@@ -115,10 +123,11 @@ func (s *Service) allocateTasks(runID string, actions []*Action) ([]*Task, error
 				return nil, err
 			}
 			tasks = append(tasks, &Task{
-				RunID:  runID,
-				Action: action,
-				HookID: hook.ID,
-				Hook:   h,
+				RunID:     runID,
+				HookRunID: NewRunID(),
```
Comment (itaiad200) [inline actions/service.go:127]: We have a runID for each hook run now?

File: swagger.yml [hunk 21]
```diff
@@ -2319,7 +2325,7 @@ paths:
           schema:
             $ref: "#/definitions/error"
 
-  /repositories/{repository}/actions/runs/{run_id}/hooks/{hook_id}/output:
+  /repositories/{repository}/actions/runs/{run_id}/hooks/{hook_run_id}/output:
```
Comment (itaiad200) [inline swagger.yml:2328]: Providing another nested generated ID is less convenient for the user. Isn't `hook_run_id` sufficient for us to fetch the relevant data?

File: actions/db_run_result_iterator.go (hunk 18 above)
Comment (nopcoder) [reply actions/db_run_result_iterator.go:60]: right, it made sense when we supported seek as we like to have >= for the first call and > for the rest.
will remove the use of this state

**Comment by nopcoder** [comment]
Replaced by PR #1486

**Pull request closed without merging**